src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentBehindCodeActionProvider.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Components;
+using Microsoft.AspNetCore.Razor.Language.Extensions;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+using Microsoft.AspNetCore.Razor.Threading;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Razor;
+internal sealed class ExtractToComponentBehindCodeActionProvider : IRazorCodeActionProvider
+{
+    private readonly ILogger _logger;
+
+    public ExtractToComponentBehindCodeActionProvider(ILoggerFactory loggerFactory)
+    {
+        if (loggerFactory is null)
+        {
+            throw new ArgumentNullException(nameof(loggerFactory));
+        }
+
+        _logger = loggerFactory.GetOrCreateLogger<ExtractToComponentBehindCodeActionProvider>();
+    }
+    public Task<IReadOnlyList<RazorVSInternalCodeAction>?> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+    {
+        if (context is null)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        if (!context.SupportsFileCreation)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        if (!FileKinds.IsComponent(context.CodeDocument.GetFileKind()))
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var syntaxTree = context.CodeDocument.GetSyntaxTree();
+        if (syntaxTree?.Root is null)
+        {
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var owner = syntaxTree.Root.FindInnermostNode(context.Location.AbsoluteIndex, true);
+        if (owner is null)
+        {
+            _logger.LogWarning($"Owner should never be null.");
+            return SpecializedTasks.Null<IReadOnlyList<RazorVSInternalCodeAction>>();
+        }
+
+        var componentNode = owner?.FirstAncestorOrSelf<MarkupElementSyntax>();
```

Comment:
Nit: Can remove the `?` here. On line 58 we've verified that `owner` can't be null, so not only is it unnecessary, it actually sends a message to the compiler that we're not sure about it, even though we are, and could cause the compiler to produce more warnings for subsequent uses of `owner`.